devdoc/Metrics.rst

```diff
@@ -0,0 +1,282 @@
+
+
+ESP
+===
+
+There are some interesting decisions with respect to ESP and collection of metrics. Different applications within ESP present different use cases for collection. Ownership of a given task drives some of these use cases. Take workunit queues. If ownership of the task, with respect to metrics, is WsWorkunits, then uses cases are centric to that component. However, if agents listening on the queue are to report metrics, then a different set of use cases emerge. It is clear that additional work is needed to generate clear ownership of metrics gathered by ESP and/or the tasks it performs.
+
+ESP needs to report the *activeTransactions* value from the TxSummary class(es). This gives an indication of how busy the ESP is in terms of client requests.
+
+Direct measurement of response time in requests may not be useful since the type of request causes different execution paths within ESP that are expected to take widely varying amounts of time. Creation of metrics for each method is not recommended. However, two possible solutions are to a) create a metric for request types, or b) use a histogram to measure response time ranges. Another option mentioned redefines the meaning of a bucket in a histogram. Instead of a numeric distribution, each bucket represents a unique subtask within an overall "metric" representing a measured operation. This should be explored whether for operational or developmental purposes.
+
+For tracking specific queries and their health, the feeling is that logging can accomplish this better than metrics since the list of queries to monitor will vary between clusters. Additionally, operational metrics solving the cases mentioned above will give a view into the overall health of ESP which will affect the execution of queries. Depending on actions taken by these metrics, scaling may solve overload conditions to keep cluster responsiveness acceptable. Internal problems with queries should be diagnosed from deep drill down logging.
+
+For Roxie a workunit operates as a service. Measuring service performance using a histogram to capture response times as a distribution may be appropriate. Extracting the 95th percentile of response time may be useful as well.
+
+There does not be a use case requiring consistency between values of different metrics
+
+
+Dali Use Cases
+==============
+
+From information gathered, Dali plans to keep counts and rates for many of the items it manages.
+
+**********************
+Framework Requirements
+**********************
+Not sure if we need this section or not. Might be good to collect high level instead of per area requirements.
+
+Metric Values
+===============
+Metric value requirements.  Capture the following:
+
+* intrinsic value ownership
```

Comment:
I am not clear what this means.

Reply:
The section in which this lives is more of a place holder to capture any requirements we decide to specifically enumerate. The current text is suggesting items that may be covered in the section. Perhaps it should be removed until such requirements exist and we decide to document them.